### PR TITLE
ci: add ci pipeline

### PR DESCRIPTION
**Pull request by weichweich**
## related https://github.com/KILTprotocol/ticket/issues/2808

* add pipeline to lint and build the frontend and backend
   * linting the backend didn't work yet
* add a `start-prod` version to the backend that doesn't use nodemon
* remove `well-known` from build step
    * well-known is part of the configuration and should not be part of the build script
    * When we build the docker images, we don't want to include a well-known DID configuration
* remove old script from the root package.json
* update to a newer yarn version to be able to remove devDependencies from `node_modules`
    * https://github.com/yarnpkg/yarn/issues/6373#issuecomment-772480514

The pipeline is failing at the moment and should be fixed in another PR.

## Backend Image

* we use node instead of ts-node
    * any errors are revealed in the build step, not when we start the container
    * lesser dependencies which allows for smaller image size
    * supposedly more resource efficient (remains to be proven)
* We don't optimize the image to only reinstall dependencies when needed
    * there could be two copy steps. First only copy package.json files, second copy the rest of the project. This allows to have faster builds since the installed dependencies can be cached.
    * this could be an optimization for later

## Frontend Image

* even though it's only static assets, we use a container to have a simpler deployment
    * frontend and backend are deployed using the same technology which makes it easier to coordinate updates (both are updated at the same time)
* The backend needs to run on the same domain as the frontend.
    * we don't support configuring a backend URL
    * all requests are relative to the current domain

### CODE REVIEW

File: scripts/wellKnownDIDConfiguration.ts
Comment (kilted-andres) [inline scripts/wellKnownDIDConfiguration.ts:1]: These linting-change happens every time I open this file. Let's see if it stays after you commit it.

File: package.json [hunk 1]
```diff
@@ -11,16 +11,18 @@
...
+    "node": ">= v18.17.0"
+  },
   "scripts": {
     "environment": "ts-node --skipProject ./scripts/genesisEnvironmentVariables.ts",
-    "build:client": "cd ./frontend && yarn build",
-    "build:well-known": "ts-node --skipProject ./scripts/genesisWellKnownDidConfig.ts",
-    "build": "yarn build:well-known && yarn build:client",
-    "start:client": "ts-node ./scripts/updatePorts.ts && cd ./frontend && yarn start",
-    "start:server": "cd ./backend && yarn start",
+    "well-known-config": "ts-node --skipProject ./scripts/genesisWellKnownDidConfig.ts",
```
Comment (kilted-andres) [inline package.json:19]: maybe:
```suggestion
    "config:well-known": "ts-node --skipProject ./scripts/genesisWellKnownDidConfig.ts",
```

File: scripts/wellKnownDIDConfiguration.ts [hunk 2]
```diff
@@ -262,7 +263,7 @@ export async function selfAttestCredential(
 
   let submitTx: Kilt.SubmittableExtrinsic
 
-  const signCallback = async ({ data }: any) => ({
+  const signCallback = async ({ data }: { data: string | Uint8Array }) => ({
```
Comment (kilted-andres) [inline scripts/wellKnownDIDConfiguration.ts:266]: nice

File: package.json [hunk 3]
```diff
@@ -11,16 +11,18 @@
...
+    "build:frontend": "cd ./frontend && yarn build",
+    "build:server": "cd ./backend & yarn check",
+    "build": "yarn build:frontend && yarn build:server",
+    "start:frontend": "ts-node ./scripts/updatePorts.ts && cd ./frontend && yarn start",
+    "start:server": "cd ./backend && yarn dev",
```
Comment (kilted-andres) [inline package.json:24]: Isn't always the pairs: 
- either `frontend` and `backend`
-  or `server` and `client`

File: .github/workflows/pull-request.yml
Comment (kilted-andres) [inline .github/workflows/pull-request.yml:1]: I can give no input to this file.

File: scripts/wellKnownDIDConfiguration.ts
Comment (weichweich) [reply scripts/wellKnownDIDConfiguration.ts:1]: You need to config your editor then.

File: package.json (hunk 3 above)
Comment (weichweich) [reply package.json]: [d48308](https://github.com/KILTprotocol/web3-login-demo/pull/35/commits/5d48308e402818e595f47deea15ee44c6dc69b66)

File: backend/package.json [hunk 4]
```diff
@@ -6,7 +6,9 @@
   "repository": "git@github.com:KILTprotocol/web3-login-demo.git",
   "license": "BSD-4-Clause",
   "scripts": {
-    "start": "nodemon"
+    "start": "nodemon",
+    "start-prod": "node ./build/server.js",
+    "build": "npx tsc"
```
Comment (arty-name) [inline backend/package.json:11]: AFAIK npx is unnecessary for package.json scripts
```suggestion
    "build": "tsc"
```

File: backend/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+# We copy everything that we might need.
+COPY . .
+
+# we only install (dev-)dependencies for the backend
+RUN yarn workspaces focus web3-login-demo-backend
+# build the backend (node is more resource efficient than ts-node)
+RUN yarn build:backend
+# remove all devDependencies again
+RUN yarn workspaces focus web3-login-demo-backend --production
```
Comment (arty-name) [inline backend/Dockerfile:13]: If you optimize the container size you might want to add `yarn cache clean`

**Review by arty-name — APPROVED** [review]
Somehow I could not find where `yarn well-known-config` is called

**Comment by weichweich** [comment]
> Somehow I could not find where `yarn well-known-config` is called

Thanks for the review. The well known config is build before this deployed and after that injected as a configuration. https://github.com/KILTprotocol/kubernetes/blob/716d77fef4d4bdd1caf128560862d52479cfe9df/login-demo/base/frontend.yaml#L41

**Pull request merged**